### ISSUE TITLE
Run Nix setup.sh?

### ISSUE THREAD

**Issue by andychu** [oilshell/oil#26]
Hm this uses several (ugly) bash features not implemented in Oil:

We should definitely be able to parse it -- running it will take some work.

- [x] `trap` builtin
- [ ] compatibility issue: `case "$(type -t "$hookName")" in` fails in OSH because we respect `errexit` in command sub, but bash doesn't
- [ ] compatibility issue: `for i in "${nativePkgs[@]}"; do` -- OSH doesn't allow indexing empty var/string with [@]
- [x] `type` builtin (should be easy, only need `-t`)
- [x] `read` flags: -r, -n etc. (but lack of `-r` not implemented)
- [x]  `oldOps="$(shopt -po nounset)"; ...; eval "$oldOpts"` (use case for "with" in Oil)
- [x] `setup.sh` has same issue as Gentoo in issue #19, trying to run `[` with empty `$PATH`
  - [x] `test -x`
- [x] `content="${content//"$pattern"/$replacement}"`
- [x] `${FUNCNAME[@]}`
- [x] `${!varRef}`  -- completion scripts also use this
- [x] `shopt -s nullglob`
- [x] `local -a times=(...)` (unparsed)
- [x] `declare -a` (unparsed)
- [x] append e.g. `eval "$var"'+=("$pkg")'` (unparsed)
- [x] might be tricky: `if [[ -z "$makeFlags" && ! ( -n "$makefile" || -e Makefile || -e makefile || -e GNUmakefile[[ ) ]]`

Deferred:

- [ ] named file descriptors:  `exec {fd}< "$fn"` (unparsed)
- [ ] closing descriptor: `exec {fd}<&-` (unparsed)
- [x] `printf` builtin -- `%q` is used, but external command should be OK

https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/generic/setup.sh#L318

**Comment by andychu** [comment]
FYI @lheckemann

**Comment by lheckemann** [comment]
Awesome, thanks! I'll see if I can find the time to contribute some of this functionality myself.

**Comment by andychu** [comment]
OK great I'm going to tackle the parsing stuff first, so:

- `declare -a` / `local -a`
- `+=`
- `{fd}<`
- `<&-`

The `type` builtin is probably the easiest one ... the command dispatch is in `core/cmd_exec.py`.  Let me know if anything doesn't make sense, or if some more docs on the wiki would help.

**Comment by andychu** [comment]
Note on last bullet point.  This works fine:

    $ bin/osh -n -c '[[ -z "$makeFlags" && ! ( -n "$makefile" || -e Makefile || -e makefile || -e GNUmakefile[[ ) ]]'             

The LST doesn't preserve the grouping parentheses, but that's (maybe) an issue for conversion and not execution.

**Comment by andychu** [comment]
@lheckemann If you pull master, the setup.sh should parse now, with:

    osh -n setup.sh

It doesn't run yet though -- failing on the shopt builtin.

Also, I believe these lines in isELF and isScript:

    exec {fd}< "$fn"
    read -r -n 2 -u "$fd" magic
    exec {fd}<&-

should be replaced with:

    read -r -n 2 magic < "$fn"

The pattern they're using is "open to fd", then "read -u $fd", then "close $fd".  There's no point to this as you can just redirect stdin of `read` directly from the file.

(This similar to the point here [1] I made about `exec 6>&1` then `exec 1>&6 6>&-` in an Apache yetus script; it can be vastly simplified to a simple redirect of a function.  I need to write a blog post about this)

[1] https://www.reddit.com/r/oilshell/comments/6kygto/osh_runs_real_shell_programs/djr8ier/

OSH is still pretty far from running this script, so I wouldn't change it for OSH.  But if there is a decent way of testing it, maybe someone wants to change it for readability / simplicity / slight efficiency.

I'm going to put off the `exec {fd}<` and `exec <&-` constructs for last, because I still haven't seen a good use of them.

**Comment by lheckemann** [comment]
Great! I'll take a look at implementing some of the missing functionality as soon as I've managed to get oil building from the repo.

**Comment by lheckemann** [comment]
wrt printf — I noticed that on my NixOS system (with coreutils 8.26)
```
$ printf "%q\n" $'\n'
$'\n'
$ $(which printf) "%q\n" $'\n'
''$'\n'
```

Meanwhile on an old debian system (coreutils 8.23)
```
linus@soundray:~$ printf "%q\n" $'\n'
$'\n'
linus@soundray:~$ $(which printf) "%q\n" $'\n'
/usr/bin/printf: %q: invalid conversion specification
```

So depending on the coreutils version, support for `%q` will either be absent or (as far as I can tell) incorrect… But bash has had it for a long time now. I guess this is another one of those issues to handle if anyone ever needs it.

**Comment by andychu** [comment]
OK good point about `%q`.  I added a note about it above.  It looks like it's used mostly for debug output, and it's easy for OSH to print an unambiguous represnetation of a string, even if it's not exactly `%q`.

I'm not opposed to a full printf builtin, but it's a fair amount of work that probably doesn't have to block this.

**Comment by andychu** [comment]
Oh also make sure you can run say `test/spec.sh builtins` or `test/spec.sh smoke`.  That's on the Contributing page.  It's pretty essential for implementing say `type -t`.

`test/spec.sh all` does all of them in parallel, without showing output.

**Comment by Ericson2314** [comment]
My favorite bash oddity is the hardly-believable `local -a "foo=(\"\$bar\")"`.

**Comment by andychu** [comment]
Hm wow I didn't know about `local -a 'foo=("$bar")'`.  It also works with "declare".

That's very similar to issue #3.  It's basically runtime parsing of arithmetic expressions inside strings. 

Bash is very confused about data vs. code.

I think this is basically because `local foo=bar` is a hack to begin with.  OSH parses it statically, but to bash, `local` is just a builtin that gets an `argv` array, and the code is passed dynamically.

So `local foo=bar` is the same as `local 'foo=bar'`.  And then that makes local 'foo=(1 2 3)' work too.

**Comment by andychu** [comment]
https://github.com/oilshell/oil/commit/a5c107a34540a8340dc1fbe29327664bad522463

`$FUNCNAME` tested but not implemented

**Comment by Ericson2314** [comment]
@andychu I just learned it is *array + `(`* specific:

```bash
$ declare -a 'foo="$bar"'; set | grep foo
_='foo="$bar"'
foo=([0]="\"\$bar\"")
```
```bash
declare +a 'foo=("$bar")'; set | grep foo
_='foo=("$bar")'
foo='("$bar")'
```

**Comment by andychu** [comment]
Interesting examples.  I haven't encountered stuff like this in the wild, but I think I get what's going on there, if not *why*.

1. `-a` means array, so the `declare` builtin dynamically parses its first string arg as an assignment.  So you get an array of one element.  `$bar` is expanded to the empty string.
2. `-a` means array, so `declare` turns everything to the right of `=` into a string, and then puts that into an array of one element.  It doesn't expand `$bar`.  This is weird but that's what it's doing.
3. `+a` means it's not an array, so it turns everything to the right of `=` into a string.  It doesn't expand `$bar`.

This would make a great post for "Bash: the Awful Parts" but I'm intentionally concentrating on OSH/Oil over the blog.  But if I get Oil done and need to justify it, there is a great example :)

Right now OSH doesn't have an array attribute of a cell/location -- it only has Str and StrArray values.  So the type is carried along with the value rather than the location, which I think makes more sense.  It remains to be seen whether this breaks any bash programs, but I think some of these semantics.

The problem is that bash has two ways to express an array -- `-a` vs `+a`, and `(a b c)` vs `'a b c'`.  I don't see why these two ways are necessary.  In Oil I just use the presence of array literals to tell me.

That is, `declare -a myarray` can simply be written `declare myarray=()`.

This doesn't work in bash but it has no ambiguity.  It's a little more Python-like, where objects have types and not variables.

**Comment by Ericson2314** [comment]
The type of an undefined variable is observable here:
```bash
$ declare -i a
$ a+='   1'
$ echo $a
1
$ a+='   1'
$ echo $a
2
```
I fear bash is poorly designed to the point where almost all internal implementation details are not abstracted way.

**Comment by andychu** [comment]
Hm that's another good one.  And note that `(( a += '  1' ))` does something different -- it's always arithmetic, even without the `-i` attribute.

I hope that OSH doesn't have to implement to make real programs run, but I saved them here:

https://github.com/oilshell/oil/blob/master/spec/type-compat.test.sh

Nix is definitely using some nooks and crannies of bash, but nothing too insane.  The eval / var ref stuff is a little odd but I think it should work in OSH.

I understand `{!varRef}` but `declare -n` is pretty wierd:

https://github.com/oilshell/oil/blob/master/spec/var-ref.test.sh

If you have any slightly easier bash scripts to run I would be interested in those too :)

**Comment by andychu** [comment]
FYI I turned the `exec` examples into a blog post, along with the other example I mentioned above:

http://www.oilshell.org/blog/2017/08/12.html

tl;dr isELF and isScript in `setup.sh` can be written in a simpler fashion, like this:

https://github.com/oilshell/blog-code/blob/master/hard-coded-descriptors/demo.sh#L24

**Comment by lheckemann** [comment]
Re: `printf` again, looking at it again I see that the output from coreutils, while awkward, is correct after all. So not implementing it as a builtin should be fine as long as we're ok with depending on a recent version of coreutils.

**Comment by Ericson2314** [comment]
@andychu I rewrote `is*` that more complicated way as I thought I was gong to call them from a context when stdin was already in use (`while read; do .. done < stuff`).

**Comment by andychu** [comment]
@Ericson2314 There is no problem with using `isElfSimple` while `stdin` is being redirected, because the `<` redirect implicitly saves and restores process state.

The way `read < foo` works is:

1. save process descriptor state, open file `foo`, make it descriptor 0
2. execute `read` builtin, which uses descriptor 0
3. restore process descriptor state.  Now descriptor 0 is whatever it was before.

Nothing but `read` happens between steps 1 and 3.  If there were any concurrency, it would happen in a *different* process, so it's always safe.

I think this wasn't have been entirely clear to me before writing a shell, but it's clear now.

Demo here:

https://github.com/oilshell/blog-code/commit/f61ff5cdc9b2da85ccda54a49a4e94a1932f83e3

```
$ ./demo.sh isElfSimpleWithStdin 
1
YES
NO

2
YES
NO

3
YES
NO
```

**Comment by andychu** [comment]
@Ericson2314 I noticed this commit

https://github.com/NixOS/nixpkgs/commit/81194eef45e2c03018687be60c2c695a1729df36

where you noted:

> `${foo+"${foo[@]}"}` will prevent `set -u` problems with empty arrays and older without making a single '' in the empty case.

but you say this happens only on old versions of bash?  Do you have a new version of bash where it doesn't happen?  This came up on the bash mailing list:

http://lists.gnu.org/archive/html/help-bash/2017-09/threads.html

I get this undesired behavior in bash 4.3, and your workaround works.  Just curious where you found this or if you came up with it... it is not obvious to me!

```
$ declare -a empty=()  # empty array, not an unset variable
$ set -u
$ echo "${empty[@]}"
-bash: empty[@]: unbound variable
```


```
$ bash --version
GNU bash, version 4.3.48(1)-release (x86_64-pc-linux-gnu)
```

**Comment by andychu** [comment]
I released OSH 0.1 with all these bug fixes:

http://www.oilshell.org/blog/2017/09/09.html

There are a few issues left, as noted at the top, but I have some other bugs to work on.  I'd appreciate any more bug reports or feedback on what distros want from a shell.

I know it is slow and unpolished, so I'd like to add something you can't get from bash in the near term.  I was thinking of adding of adding better debugging features:

- a better `set -x`, in particular one that shows all the exit codes, so you can figure out why a script that uses `set -e` exited
- being able to dump all variables on the stack after a failure and inspect them.  I'm not going to implement a debugger but something like a 'crashdump' builtin might be useful for debugging.

(The other place it's already better than bash is `osh -n`, which does static parsing.)

I think the initial use case for OSH is building Linux distros, i.e. large shell programs (as opposed to interactive usage). 

I think I will concentrate on Alpine Linux for the next release, because it's very small, but I'm still interested in feedback from Nix maintainers and users.

(BTW I also discovered on the mailing list that bash 4.4 fixed the `set -u` and empty array bug.)

**Comment by andychu** [comment]
@lheckemann @Ericson2314 Wondering if you guys had any feedback on this?

Sorry if this issue got pinged a lot an overwhelmed your mailbox.  I'm not sure if it sends notifications when I edit the top-level comment, which I did a lot.

**Comment by andychu** [comment]
Thanks for finding that!  I just replied to the pull request.

What's the command I can reproduce this with?  I was running `osh setup.sh` or `'bin/osh -c "source osh.s"'` and getting a different error.

```
$ bin/osh -c 'source setup.sh'                                                                     
osh warning: *** trap not implemented ***
Line 346 of 'setup.sh'
  for i in "${nativePkgs[@]}"; do
            ^~
osh failed: Can't index string with @: (Str s:"")
```

**Comment by lheckemann** [comment]
Sorry, I think the failure wasn't actually on that line after all. My best guess is that it's related to the "Hack around old bash being bad and thinking empty arrays are undefined." on the lines after, which is a bit mind-bending. I'll see if I can figure out a minimal test case for it.

**Comment by lheckemann** [comment]
Update: I'm just a bit of an idiot and mistook rebuilding a bunch of stuff for everything working. The same error still occurs with the `: $hookName`.

**Comment by andychu** [comment]
@lheckemann I didn't follow that last part, but please try again after the fix I just made and let me know what the current error is.

- I can explain the empty arrays hack if you are interested... I just learned about that and it took me awhile to figure out.  It is a change in bash 4.4 -- 4.3 treats empty arrays with set -u differently.  (So it's not actually a very old bash, 4.4 came out last year)
- If you are still getting this, file a bug with the script /command that triggers it and I can probably fix it.

```
Line 0 of '<unknown>'
  <no position info for token>
osh failed: Undefined variable
```

**Comment by andychu** [comment]
TODO: Try building "hello" package with Nix as described here:

https://www.reddit.com/r/oilshell/comments/7qn0p8/success_with_aboriginal_alpine_and_debian/dsrnp0q/

Now that I look at this bug again, the `errexit` compatibility issue I mentioned was fixed due to Alpine.  As far as I remember, `setup.sh` used some pretty hairy constructs but it's probably worth another try.

**Comment by andychu** [comment]
I checked off a few more things on this dormant issue.  We're now using this page to prioritize:

https://github.com/oilshell/oil/wiki/Shell-Programs-That-Run-Under-OSH
